lib/src/main/kotlin/StateActionNode.kt

```diff
@@ -1,17 +1,42 @@
-internal class StateActionNode<TAction>(
-        parent: NodeBase?,
-        val parentAction: TAction?) : NodeBase(parent) {
+class StatelessActionNode<TState, TAction>(
+        parent: StatelessActionNode<TState, TAction>?,
+        inducingAction: TAction?) : Node<TAction, StatelessActionNode<TState, TAction>>(parent, inducingAction) {
 
-    var children = mutableListOf<StateActionNode<TAction>>()
+    private var _state: TState? = null
+    private var _validActions: Iterable<TAction>? = null
 
-    fun parentStateAction() : StateActionNode<TAction>? {
-        return if (parent == null)
-            null
-        else
-            parent as StateActionNode<TAction>
+    var state: TState
+        get() {
```

Comment:
@JunTaoLuo it is throwing an error now

```
java.lang.Exception: Simulation not run
	at StatelessActionNode.getState(StateActionNode.kt:10)
	at StatelessSolver.expandNode(StatelessSolver.kt:44)
	at StatelessSolver.expandNode(StatelessSolver.kt:3)
	at SolverBase.iterateStep(SolverBase.kt:44)
	at SolverBase.constructTree(SolverBase.kt:22)
```
with the new changes, mind explaining the rationale behind these changes once more?

Reply:
Its returning a `null` where its not suppose to when running StatelessSolver

```
override fun expandNode(node: StatelessActionNode<TState, TAction>): StatelessActionNode<TState, TAction> {
        // If the node is terminal, return it
        if (mdp.isTerminal(node.state)) {
            return node
        }
  ```

Reply:
Details on the error

```
> Task :GridWorldMainKt.main() FAILED
Exception in thread "main" java.lang.Exception: Simulation not run
	at StatelessActionNode.getState(StateActionNode.kt:10)
	at StatelessSolver.expandNode(StatelessSolver.kt:44)
	at StatelessSolver.expandNode(StatelessSolver.kt:3)
	at SolverBase.iterateStep(SolverBase.kt:44)
	at SolverBase.constructTree(SolverBase.kt:22)
	at GridWorld.GridWorldGridSolve.getWorldSolve(GridWorldGridSolve.kt:45)
	at GridWorldMainKt.main(GridWorldMain.kt:22)
	at GridWorldMainKt.main(GridWorldMain.kt)
```
The new solver is failing to expandNode
```
var state: TState
        get() {
            return _state ?: throw Exception("Simulation not run")
        }
        set(value) {
            _state = value
        }
 ```

Reply:
Ah yea, I meant to fix it, but ran out of time. I've been rather busy.